lib/appsignal/check_in/debouncer.rb

```diff
@@ -0,0 +1,190 @@
+# frozen_string_literal: true
+
+module Appsignal
+  module CheckIn
+    class Debouncer
+      SHORT_DEBOUNCE_SECONDS = 0.1
+      LONG_DEBOUNCE_SECONDS = 10
+
+      def initialize
+        # The mutex is used to synchronize access to the events array, the
+        # waker thread and the main thread, as well as queue writes
+        # (which depend on the events array) and closes (so they do not
+        # happen at the same time that an event is added to the debouncer)
+        @mutex = Mutex.new
+        # The thread will be started when an event is first added.
+        @thread = nil
+        @queue = Thread::Queue.new
+        @events = []
+        # The waker thread is used to schedule debounces. It will be started
+        # when an event is first added.
+        @waker = nil
+        # For internal testing purposes.
+        @transmitted = 0
+      end
+
+      def send(event)
+        unless Appsignal.active?
+          Appsignal.internal_logger.debug(
+            "Cannot transmit #{describe([event])}: AppSignal is not active"
+          )
+          return
+        end
+
+        @mutex.synchronize do
+          if @queue.closed?
+            Appsignal.internal_logger.debug(
+              "Cannot transmit #{describe([event])}: AppSignal is stopped"
+            )
+            return
+          end
+          add_event(event)
+          # If we're not already waiting to be awakened from a scheduled
+          # debounce, schedule a short debounce, which will push the events
+          # to the queue and schedule a long debounce.
+          schedule_waker(SHORT_DEBOUNCE_SECONDS) if @waker.nil?
+
+          Appsignal.internal_logger.debug(
+            "Scheduling #{describe([event])} to be transmitted"
+          )
+
+          # Make sure to start the thread after an event has been added.
+          @thread ||= Thread.new(&method(:run))
+        end
+      end
+
+      def stop
+        @mutex.synchronize do
+          # Flush all events before closing the queue.
+          push_events
+        rescue ClosedQueueError
+          # The queue is already closed (by a previous call to `#stop`)
+          # so it is not possible to push events to it anymore.
+        ensure
+          # Ensure calling `#send` after `#stop` raises a `ClosedQueueError`,
+          # by killing the waker thread, which will cause `#push_events` to
+          # enqueue the events immediately, and closing the queue, which will
+          # cause the queue to raise a `ClosedQueueError`.
+          kill_waker
+          @queue.close
+
+          # Block until the thread has finished.
+          @thread&.join
+        end
+      end
+
+      # @api private
+      # For internal testing purposes.
+      attr_reader :thread, :waker, :queue, :events, :transmitted
+
+      private
+
+      def run
+        loop do
+          events = @queue.pop
```

Comment:
Won't this transmit a single event? (I think `queue.pop` only pops one. value)

Reply:
You're correct that `queue.pop` only pops one value, but [the value it pops is an array of events](https://github.com/appsignal/appsignal-ruby/pull/1245/files/97e0cf6ac48b61e37ae89aba801c5b9d885714a6#diff-777a84febbe13fff07d37faf6813fbeb192a5011bdb4622d808ef5204e731426R180-R183) to be sent.